File: desktop-mend/src/main/java/co/samco/mend4/desktop/dao/impl/SettingsImpl.java
package co.samco.mend4.desktop.dao.impl;

import co.samco.mend4.core.OSDao;
import co.samco.mend4.core.Settings;
import co.samco.mend4.core.exception.CorruptSettingsException;

import java.io.*;
import java.util.Properties;

public class SettingsImpl implements Settings {
    private final OSDao osDao;
    private final File settingsFile;
    private Properties _propertiesCache;
    private final String corruptSettingsExceptionText;
    private boolean propertiesNotFound = false;

    public SettingsImpl(OSDao osDao, File settingsFile, String corruptSettingsExceptionText) {
        this.osDao = osDao;
        this.settingsFile = settingsFile;
        this.corruptSettingsExceptionText = corruptSettingsExceptionText;
    }

    private Properties getProperties() throws IOException {
        if (_propertiesCache == null) {
            _propertiesCache = new Properties();
            try (InputStream fis = osDao.getInputStreamForFile(settingsFile)) {
                _propertiesCache.loadFromXML(fis);
            } catch (FileNotFoundException e) {
                propertiesNotFound = true;
            }
        }
        return _propertiesCache;
    }

    private void saveProperties(Properties properties) throws IOException {
        _propertiesCache = properties;
        try (OutputStream fos = osDao.getOutputStreamForFile(settingsFile)) {
            properties.storeToXML(fos, "");
        } catch (IOException e) {
            throw e;
        }
    }

    @Override
    public void setValue(Name name, String value) throws IOException {
        Properties properties = getProperties();
        properties.setProperty(name.toString(), value);
        saveProperties(properties);
    }

    @Override
    public boolean valueSet(Name name) throws IOException {
        if (propertiesNotFound) {
            return false;
        }
        Properties properties = getProperties();
        String value = properties.getProperty(name.toString());
        return value != null;
    }

    @Override
    public String getValue(Name name) throws IOException, CorruptSettingsException {
        Properties properties = getProperties();
        String value = properties.getProperty(name.toString());
        if (value == null) {
            throw new CorruptSettingsException(corruptSettingsExceptionText, name.toString());
        }
        return value;
    }
}

File: desktop-mend/src/main/java/co/samco/mend4/desktop/dao/impl/SettingsImpl.kt
package co.samco.mend4.desktop.dao.impl

import co.samco.mend4.core.Settings
import co.samco.mend4.desktop.dao.OSDao
import co.samco.mend4.desktop.dao.SettingsDao
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.Properties
import javax.inject.Inject

class SettingsImpl @Inject constructor(osDao: OSDao) : SettingsDao {
    private val settingsFile = osDao.getSettingsFile()

    private var _propertiesCache: Properties? = null

    private val properties: Properties
        get() {
            if (_propertiesCache == null) {
                _propertiesCache = Properties()
                if (settingsFile.exists()) FileInputStream(settingsFile).use {
                    _propertiesCache!!.loadFromXML(it)
                }
            }
            return _propertiesCache!!
        }


    private fun saveProperties(properties: Properties) {
        _propertiesCache = properties
        FileOutputStream(settingsFile).use { properties.storeToXML(it, "") }
    }

    override fun firstNonExistent(required: Array<Settings.Name>): Settings.Name? =
        required.firstOrNull { properties.getProperty(it.encodedName) == null }

    override fun setValue(name: Settings.Name, value: String) {
        properties.setProperty(name.encodedName, value)
        saveProperties(properties)
    }

    override fun getValue(name: Settings.Name): String? =
        properties.getProperty(name.encodedName)
}